### ISSUE TITLE
'add-config' and 'remove-config' does not work as expected

### ISSUE THREAD

**Issue by Jaylimboonkiat** [pingidentity/pingidentity-devops-getting-started#360]
**Describe the bug**
When performing 'remove-config' as stated in the documentation
```
  "remove-config":[
  	{
  	  "key": "id",
	  "value": "ProvisionerDS"
  	}
  ]
```
The configuration still remains in the output data.json.subst.

When performing 'add-config' as stated in the documentation
```
  "add-config":[
	  {
	    "resourceType": "/idp/spConnections",
	    "item":
		{
                    "name": "httpbin3.org",
                    "active": false,
		    ...
		}
	  }
  ]
```
The configuration is not added in the output data.json.subst.

**Expected behavior**
When performing 'remove-config', all configurations with expected key-value pair should be removed.
When performing 'add-config', configuration with new resourceType should be added.

**Environment:**
 - Github Repo: https://github.com/pingidentity/pingidentity-devops-getting-started/tree/master/99-helper-scripts/ping-bulkconfigtool

**Comment by samir-gandhi** [comment]
Could you provide more details: 
1. your config.json
2. sample data.json (secret text removed)
3. the command you used to run the tool.

**Comment by Jaylimboonkiat** [comment]
Sample data.json
```{
  "metadata": {
    "pfVersion": "11.0.2.0"
  },
  "operations": [
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "com.pingidentity.crypto.LunaJCEManager",
      "items": [
        {
          "id": "enabled-cipher-suites",
          "type": "MAP",
          "mapValue": {
            "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA": "",
            "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA": "",
            "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384": "",
            "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA": "",
            "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384": "",
            "TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA": "",
            "TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA": "",
            "TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA384": "",
            "TLS_ECDH_RSA_WITH_AES_128_CBC_SHA": "",
            "TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDH_RSA_WITH_AES_256_CBC_SHA": "",
            "TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384": "",
            "TLS_EMPTY_RENEGOTIATION_INFO_SCSV": "",
            "TLS_RSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_RSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_RSA_WITH_AES_256_CBC_SHA": "",
            "TLS_RSA_WITH_AES_256_CBC_SHA256": ""
          }
        }
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "com.pingidentity.crypto.NcipherJCEManager",
      "items": [
        {
          "id": "enabled-cipher-suites",
          "type": "MAP",
          "mapValue": {
            "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA": "",
            "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA": "",
            "TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384": "",
            "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA": "",
            "TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384": "",
            "TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA": "",
            "TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA": "",
            "TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA384": "",
            "TLS_ECDH_RSA_WITH_AES_128_CBC_SHA": "",
            "TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDH_RSA_WITH_AES_256_CBC_SHA": "",
            "TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384": "",
            "TLS_EMPTY_RENEGOTIATION_INFO_SCSV": "",
            "TLS_RSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_RSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_RSA_WITH_AES_256_CBC_SHA": "",
            "TLS_RSA_WITH_AES_256_CBC_SHA256": ""
          }
        }
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "com.pingidentity.crypto.SunJCEManager",
      "items": [
        {
          "id": "enabled-cipher-suites",
          "type": "MAP",
          "mapValue": {
            "TLS_AES_128_GCM_SHA256": "",
            "TLS_AES_256_GCM_SHA384": "",
            "TLS_CHACHA20_POLY1305_SHA256": "",
            "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA": "",
            "TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA": "",
            "TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_ECDH_RSA_WITH_AES_128_CBC_SHA": "",
            "TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256": "",
            "TLS_EMPTY_RENEGOTIATION_INFO_SCSV": "",
            "TLS_RSA_WITH_AES_128_CBC_SHA256": "",
            "TLS_RSA_WITH_AES_128_GCM_SHA256": ""
          }
        }
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "DELETE",
      "subResource": "http-request-parameter-validation",
      "itemIds": [
        "client_id"
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "org.sourceid.common.ExpressionManager",
      "items": [
        {
          "id": "evaluateExpressions",
          "type": "STRING",
          "stringValue": ""
        }
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "org.sourceid.oauth20.domain.AuthzServerManagerImpl",
      "items": [
        {
          "id": "UnidentifiedClientExtGrants",
          "type": "STRING",
          "stringValue": "false"
        },
        {
          "id": "UnidentifiedClientROCreds",
          "type": "STRING",
          "stringValue": "false"
        }
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "org.sourceid.oauth20.domain.ClientManagerXmlFileImpl",
      "items": [
        {
          "id": "MigrationComplete8.4",
          "type": "STRING",
          "stringValue": "true"
        }
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "org.sourceid.saml20.domain.mgmt.AdminUserManager",
      "items": [
        {
          "id": "isEaRoleUpdated",
          "type": "STRING",
          "stringValue": "true"
        }
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "org.sourceid.saml20.domain.mgmt.impl.PartnerCertMigrator",
      "items": [
        {
          "id": "partner.cert.migration.complete",
          "type": "STRING",
          "stringValue": "true"
        }
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "org.sourceid.saml20.metadata.partner.impl.MetadataDirectoryHybridDbImpl",
      "items": [
        {
          "id": "MigrationComplete8.4",
          "type": "STRING",
          "stringValue": "true"
        }
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "org.sourceid.saml20.service.impl.PseudonymServiceSha1Impl",
      "items": [
        {
          "id": "entropy",
          "type": "STRING",
          "stringValue": "VXBlGWDvv70="
        }
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "DELETE",
      "subResource": "org.sourceid.saml20.service.session.StoredSessionServiceImpl",
      "itemIds": [
        "EmptySessionReplayRetentionSecs"
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "DELETE",
      "subResource": "org.sourceid.servlet.filter.SimultaneousAuthnRequestCheckingFilter",
      "itemIds": [
        "browser-id-cookie-httponly-flag",
        "browser-id-cookie-secure-flag"
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "DELETE",
      "subResource": "p1-admin-service-config",
      "itemIds": [
        "AddApplicationUrl"
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "DELETE",
      "subResource": "persistent-session-cookie-config",
      "itemIds": [
        "cookie-format"
      ]
    },
    {
      "resourceType": "/configStore",
      "operationType": "SAVE",
      "subResource": "response-header-admin-config",
      "items": [
        {
          "id": "X-Frame-Options",
          "type": "MAP",
          "mapValue": {
            "include-patterns": "*",
            "value": "DENY"
          }
        }
      ]
    },
    {
      "resourceType": "/license/agreement",
      "operationType": "SAVE",
      "items": [
        {
          "licenseAgreementUrl": "https://localhost:9999/pf-admin-api/license-agreement",
          "accepted": true
        }
      ]
    },
    {
      "resourceType": "/keyPairs/sslServer",
      "operationType": "SAVE",
      "items": [
        {
          "id": "vtcm75en83g6v1r87ytm7lihi",
          "fileData": "...",
          "format": "PKCS12",
          "encryptedPassword": "..."
        }
      ]
    },
    {
      "resourceType": "/keyPairs/sslServer/settings",
      "operationType": "SAVE",
      "items": [
        {
          "runtimeServerCertRef": {
            "id": "vtcm75en83g6v1r87ytm7lihi",
            "location": "https://localhost:9999/pf-admin-api/v1/keyPairs/sslServer/vtcm75en83g6v1r87ytm7lihi"
          },
          "adminConsoleCertRef": {
            "id": "vtcm75en83g6v1r87ytm7lihi",
            "location": "https://localhost:9999/pf-admin-api/v1/keyPairs/sslServer/vtcm75en83g6v1r87ytm7lihi"
          },
          "activeRuntimeServerCerts": [
            {
              "id": "vtcm75en83g6v1r87ytm7lihi",
              "location": "https://localhost:9999/pf-admin-api/v1/keyPairs/sslServer/vtcm75en83g6v1r87ytm7lihi"
            }
          ],
          "activeAdminConsoleCerts": [
            {
              "id": "vtcm75en83g6v1r87ytm7lihi",
              "location": "https://localhost:9999/pf-admin-api/v1/keyPairs/sslServer/vtcm75en83g6v1r87ytm7lihi"
            }
          ]
        }
      ]
    },
    {
      "resourceType": "/keyPairs/oauthOpenIdConnect",
      "operationType": "SAVE",
      "items": [
        {
          "staticJwksEnabled": false
        }
      ]
    },
    {
      "resourceType": "/dataStores",
      "operationType": "SAVE",
      "items": [
        {
          "type": "JDBC",
          "id": "ProvisionerDS",
          "maskAttributeValues": false,
          "connectionUrl": "jdbc:hsqldb:${pf.server.data.dir}${/}hypersonic${/}ProvisionerDefaultDB;hsqldb.lock_file=false",
          "driverClass": "org.hsqldb.jdbcDriver",
          "userName": "sa",
          "encryptedPassword": "...",
          "allowMultiValueAttributes": false,
          "name": "ProvisionerDS (sa)",
          "connectionUrlTags": [
            {
              "connectionUrl": "jdbc:hsqldb:${pf.server.data.dir}${/}hypersonic${/}ProvisionerDefaultDB;hsqldb.lock_file=false",
              "defaultSource": true
            }
          ]
        }
      ]
    },
    {
      "resourceType": "/notificationPublishers/settings",
      "operationType": "SAVE",
      "items": [
        {}
      ]
    },
    {
      "resourceType": "/serverSettings",
      "operationType": "SAVE",
      "items": [
        {
          "contactInfo": {},
          "rolesAndProtocols": {
            "oauthRole": {
              "enableOauth": true,
              "enableOpenIdConnect": true
            },
            "idpRole": {
              "enable": true,
              "enableSaml11": true,
              "enableSaml10": true,
              "enableWsFed": true,
              "enableWsTrust": true,
              "saml20Profile": {
                "enable": true
              },
              "enableOutboundProvisioning": true
            },
            "spRole": {
              "enable": true,
              "enableSaml11": true,
              "enableSaml10": true,
              "enableWsFed": true,
              "enableWsTrust": true,
              "saml20Profile": {
                "enable": true,
                "enableXASP": true
              },
              "enableInboundProvisioning": true,
              "enableOpenIDConnect": true
            },
            "enableIdpDiscovery": true
          },
          "federationInfo": {
            "baseUrl": "https://localhost:9031",
            "saml2EntityId": "evaluation",
            "saml1xIssuerId": "",
            "saml1xSourceId": "",
            "wsfedRealm": ""
          },
          "notifications": {
            "notifyAdminUserPasswordChanges": false
          }
        }
      ]
    },
    {
      "resourceType": "/serverSettings/systemKeys",
      "operationType": "SAVE",
      "items": [
        {
          "current": {
            "creationDate": "2019-03-01T21:13:15.312Z",
            "encryptedKeyData": "..."
          },
          "pending": {
            "creationDate": "2019-03-01T21:13:15.314Z",
            "encryptedKeyData": "..."
          }
        }
      ]
    },
    {
      "resourceType": "/serverSettings/outboundProvisioning",
      "operationType": "SAVE",
      "items": [
        {
          "dataStoreRef": {
            "id": "ProvisionerDS",
            "location": "https://localhost:9999/pf-admin-api/v1/dataStores/ProvisionerDS"
          },
          "synchronizationFrequency": 60
        }
      ]
    },
    {
      "resourceType": "/administrativeAccounts",
      "operationType": "SAVE",
      "items": [
        {
          "username": "Administrator",
          "encryptedPassword": "...",
          "description": "Initial administrator user.",
          "auditor": false,
          "active": true,
          "roles": [
            "EXPRESSION_ADMINISTRATOR",
            "ADMINISTRATOR",
            "USER_ADMINISTRATOR",
            "CRYPTO_ADMINISTRATOR"
          ]
        }
      ]
    },
    {
      "resourceType": "/certificates/revocation/settings",
      "operationType": "SAVE",
      "items": [
        {
          "crlSettings": {
            "treatNonRetrievableCrlAsRevoked": false,
            "verifyCrlSignature": true,
            "nextRetryMinsWhenResolveFailed": 1440,
            "nextRetryMinsWhenNextUpdateInPast": 60
          }
        }
      ]
    },
    {
      "resourceType": "/virtualHostNames",
      "operationType": "SAVE",
      "items": [
        {
          "virtualHostNames": []
        }
      ]
    },
    {
      "resourceType": "/redirectValidation",
      "operationType": "SAVE",
      "items": [
        {
          "redirectValidationLocalSettings": {
            "enableTargetResourceValidationForSSO": true,
            "enableTargetResourceValidationForSLO": true,
            "enableTargetResourceValidationForIdpDiscovery": true,
            "enableInErrorResourceValidation": true,
            "whiteList": []
          },
          "redirectValidationPartnerSettings": {
            "enableWreplyValidationSLO": true
          }
        }
      ]
    },
    {
      "resourceType": "/incomingProxySettings",
      "operationType": "SAVE",
      "items": [
        {
          "proxyTerminatesHttpsConns": false
        }
      ]
    },
    {
      "resourceType": "/protocolMetadata/lifetimeSettings",
      "operationType": "SAVE",
      "items": [
        {
          "cacheDuration": 1440,
          "reloadDelay": 1440
        }
      ]
    },
    {
      "resourceType": "/protocolMetadata/signingSettings",
      "operationType": "SAVE",
      "items": [
        {}
      ]
    },
    {
      "resourceType": "/serviceAuthentication",
      "operationType": "SAVE",
      "items": [
        {}
      ]
    },
    {
      "resourceType": "/oauth/accessTokenManagers/settings",
      "operationType": "SAVE",
      "items": [
        {}
      ]
    },
    {
      "resourceType": "/kerberos/realms/settings",
      "operationType": "SAVE",
      "items": [
        {
          "forceTcp": false,
          "debugLogOutput": false,
          "keySetRetentionPeriodMins": 610
        }
      ]
    },
    {
      "resourceType": "/idp/defaultUrls",
      "operationType": "SAVE",
      "items": [
        {
          "confirmIdpSlo": false,
          "idpErrorMsg": "errorDetail.idpSsoFailure"
        }
      ]
    },
    {
      "resourceType": "/sp/defaultUrls",
      "operationType": "SAVE",
      "items": [
        {
          "confirmSlo": false
        }
      ]
    },
    {
      "resourceType": "/session/settings",
      "operationType": "SAVE",
      "items": [
        {
          "trackAdapterSessionsForLogout": false,
          "revokeUserSessionOnLogout": true,
          "sessionRevocationLifetime": 490
        }
      ]
    },
    {
      "resourceType": "/session/applicationSessionPolicy",
      "operationType": "SAVE",
      "items": [
        {
          "idleTimeoutMins": 60,
          "maxTimeoutMins": 480
        }
      ]
    },
    {
      "resourceType": "/session/authenticationSessionPolicies/global",
      "operationType": "SAVE",
      "items": [
        {
          "enableSessions": false,
          "persistentSessions": false,
          "hashUniqueUserKeyAttribute": false,
          "idleTimeoutMins": 60,
          "idleTimeoutDisplayUnit": "MINUTES",
          "maxTimeoutMins": 480,
          "maxTimeoutDisplayUnit": "MINUTES"
        }
      ]
    },
    {
      "resourceType": "/oauth/authServerSettings",
      "operationType": "SAVE",
      "items": [
        {
          "defaultScopeDescription": "",
          "scopes": [],
          "scopeGroups": [],
          "exclusiveScopes": [],
          "exclusiveScopeGroups": [],
          "authorizationCodeTimeout": 60,
          "authorizationCodeEntropy": 30,
          "disallowPlainPKCE": false,
          "includeIssuerInAuthorizationResponse": false,
          "persistentGrantLifetime": -1,
          "persistentGrantLifetimeUnit": "DAYS",
          "persistentGrantIdleTimeout": 30,
          "persistentGrantIdleTimeoutTimeUnit": "DAYS",
          "refreshTokenLength": 51,
          "rollRefreshTokenValues": false,
          "refreshTokenRollingGracePeriod": 0,
          "refreshRollingInterval": 0,
          "persistentGrantReuseGrantTypes": [
            "IMPLICIT"
          ],
          "persistentGrantContract": {
            "extendedAttributes": [],
            "coreAttributes": [
              {
                "name": "USER_KEY"
              },
              {
                "name": "USER_NAME"
              }
            ]
          },
          "bypassAuthorizationForApprovedGrants": false,
          "allowUnidentifiedClientROCreds": false,
          "allowUnidentifiedClientExtensionGrants": false,
          "tokenEndpointBaseUrl": "",
          "userAuthorizationUrl": "",
          "registeredAuthorizationPath": "",
          "pendingAuthorizationTimeout": 600,
          "bypassActivationCodeConfirmation": false,
          "devicePollingInterval": 5,
          "userAuthorizationConsentPageSetting": "INTERNAL",
          "atmIdForOAuthGrantManagement": "",
          "scopeForOAuthGrantManagement": "",
          "allowedOrigins": [],
          "trackUserSessionsForLogout": false,
          "parReferenceTimeout": 60,
          "parReferenceLength": 24,
          "parStatus": "ENABLED"
        }
      ]
    },
    {
      "resourceType": "/oauth/cibaServerPolicy/settings",
      "operationType": "SAVE",
      "items": [
        {}
      ]
    },
    {
      "resourceType": "/oauth/openIdConnect/settings",
      "operationType": "SAVE",
      "items": [
        {
          "sessionSettings": {
            "trackUserSessionsForLogout": false,
            "revokeUserSessionOnLogout": true,
            "sessionRevocationLifetime": 490
          }
        }
      ]
    },
    {
      "resourceType": "/oauth/tokenExchange/processor/settings",
      "operationType": "SAVE",
      "items": [
        {}
      ]
    },
    {
      "resourceType": "/oauth/tokenExchange/generator/settings",
      "operationType": "SAVE",
      "items": [
        {}
      ]
    },
    {
      "resourceType": "/oauth/clientSettings",
      "operationType": "SAVE",
      "items": [
        {
          "clientMetadata": []
        }
      ]
    },
    {
      "resourceType": "/authenticationApi/settings",
      "operationType": "SAVE",
      "items": [
        {
          "apiEnabled": false,
          "enableApiDescriptions": true,
          "restrictAccessToRedirectlessMode": false,
          "includeRequestContext": false
        }
      ]
    },
    {
      "resourceType": "/authenticationPolicies/default",
      "operationType": "SAVE",
      "items": [
        {
          "failIfNoSelection": false,
          "authnSelectionTrees": [],
          "defaultAuthenticationSources": [],
          "trackedHttpParameters": []
        }
      ]
    },
    {
      "resourceType": "/authenticationPolicies/settings",
      "operationType": "SAVE",
      "items": [
        {
          "enableIdpAuthnSelection": false,
          "enableSpAuthnSelection": false
        }
      ]
    }
  ]
}```

Used config.json
```
{
  "config-aliases": [
    {
      "config-names": [
        "administrativeAccounts_items_Administrator_password"
      ],
      "replace-name": "PING_IDENTITY_PASSWORD",
      "is-apply-envfile": false
    }
  ],
  "remove-config": [
    {
      "key": "id",
      "value": "ProvisionerDS"
    }
  ],
  "add-config": [
    {
      "resourceType": "/dataStores",
      "item": {
        "type": "JDBC",
        "id": "JDBC-0B770A97F13C05CCB83E2F31EFDBC894B1684E62",
        "maskAttributeValues": false,
        "connectionUrl": "...",
        "driverClass": "com.mysql.jdbc.Driver",
        "userName": "...",
        "password": "...",
        "validateConnectionSql": "select 1",
        "allowMultiValueAttributes": true,
        "name": "...",
        "connectionUrlTags": [
          {
            "connectionUrl": "...",
            "tags": "direct",
            "defaultSource": true
          }
        ],
        "minPoolSize": 10,
        "maxPoolSize": 100,
        "blockingTimeout": 5000,
        "idleTimeout": 5
      }
    },
    {
      "resourceType": "/idp/adapters",
      "item": {
        "id": "agentlessadapter",
        "name": "Agentless Adapter",
        "pluginDescriptorRef": {
          "id": "com.pingidentity.pf.adapters.referenceid.IdpBackchannelReferenceAuthnAdapter",
          "location": "https://localhost:9999/pf-admin-api/v1/idp/adapters/descriptors/com.pingidentity.pf.adapters.referenceid.IdpBackchannelReferenceAuthnAdapter"
        },
        "configuration": {
          "tables": [],
          "fields": [
            {
              "name": "Authentication Endpoint",
              "value": "http://localhost:9081/v1/authenticate"
            },
            {
              "name": "User Name",
              "value": "..."
            },
            {
              "name": "Pass Phrase",
              "encryptedValue": "..."
            },
            {
              "name": "Allowed Subject DN",
              "value": ""
            },
            {
              "name": "Allowed Issuer DN",
              "value": ""
            },
            {
              "name": "Logout Service Endpoint",
              "value": ""
            },
            {
              "name": "Logout Service Mode",
              "value": "NONE"
            },
            {
              "name": "Prefix Referenced Attributes",
              "value": "false"
            },
            {
              "name": "Ignore Untracked HTTP Parameters",
              "value": "true"
            },
            {
              "name": "Send Request Parameters",
              "value": "NONE"
            },
            {
              "name": "Transport Mode",
              "value": "2"
            },
            {
              "name": "Reference Duration",
              "value": "60"
            },
            {
              "name": "Reference Length",
              "value": "30"
            },
            {
              "name": "Require SSL/TLS",
              "value": "false"
            },
            {
              "name": "Outgoing Attribute Format",
              "value": "JSON"
            },
            {
              "name": "Incoming Attribute Format",
              "value": "JSON"
            },
            {
              "name": "Skip Host Name Validation ",
              "value": "false"
            },
            {
              "name": "Relax Password Requirements",
              "value": "false"
            }
          ]
        },
        "attributeContract": {
          "coreAttributes": [
            {
              "name": "subject",
              "masked": false,
              "pseudonym": false
            }
          ],
          "extendedAttributes": [
            {
              "name": "user_email",
              "masked": false,
              "pseudonym": false
            },
            {
              "name": "analytics_id",
              "masked": false,
              "pseudonym": false
            },
            {
              "name": "last_name",
              "masked": false,
              "pseudonym": false
            },
            {
              "name": "ObjectGUID",
              "masked": false,
              "pseudonym": true
            },
            {
              "name": "has_prompt_error",
              "masked": false,
              "pseudonym": false
            },
            {
              "name": "first_name",
              "masked": false,
              "pseudonym": false
            }
          ],
          "uniqueUserKeyAttribute": "ObjectGUID",
          "maskOgnlValues": false
        },
        "attributeMapping": {
          "attributeSources": [],
          "attributeContractFulfillment": {
            "user_email": {
              "source": {
                "type": "ADAPTER"
              },
              "value": "user_email"
            },
            "subject": {
              "source": {
                "type": "ADAPTER"
              },
              "value": "subject"
            },
            "analytics_id": {
              "source": {
                "type": "ADAPTER"
              },
              "value": "analytics_id"
            },
            "last_name": {
              "source": {
                "type": "ADAPTER"
              },
              "value": "last_name"
            },
            "ObjectGUID": {
              "source": {
                "type": "ADAPTER"
              },
              "value": "ObjectGUID"
            },
            "has_prompt_error": {
              "source": {
                "type": "ADAPTER"
              },
              "value": "has_prompt_error"
            },
            "first_name": {
              "source": {
                "type": "ADAPTER"
              },
              "value": "first_name"
            }
          },
          "issuanceCriteria": {
            "conditionalCriteria": []
          }
        }
      }
    }
  ],
  "expose-parameters": [
    {
      "parameter-name": "fileData",
      "unique-identifiers": [
        "id"
      ]
    },
    {
      "parameter-name": "encryptedValue",
      "replace-name": "value",
      "unique-identifiers": [
        "name"
      ]
    },
    {
      "parameter-name": "encryptedPassword",
      "replace-name": "password",
      "unique-identifiers": [
        "id",
        "username"
      ]
    },
    {
      "parameter-name": "encryptedKeyData",
      "replace-name": "keyData",
      "unique-identifiers": [
        "id"
      ]
    },
    {
      "parameter-name": "encryptedSharedSecret",
      "replace-name": "sharedSecret",
      "unique-identifiers": [
        "id"
      ]
    },
    {
      "parameter-name": "encryptedSecret",
      "replace-name": "secret",
      "unique-identifiers": [
        "clientId"
      ]
    }
  ]
}
```

Command used
`docker run --rm -v $PWD/shared:/shared ping-bulkexport-tools:latest /shared/pf-config.json /shared/data.json /shared/env_vars /shared/data.json.subst > $PWD/shared/convert.log`

**Comment by ttranatping** [comment]
Hi @Jaylimboonkiat 

Thank you for your patience. We've updated the code to fix the issues and we are in the process of updating the bulk config docker image to have those changes.

The add-config had a reference defect, so it wasn't being processed.

The remove-config for ProvisioningDS succeeded for the intended node, but there is a second reference to id=ProvisioningDS in the input data JSON under outboundProvisioning.

`    {
      "resourceType": "/serverSettings/outboundProvisioning",
      "operationType": "SAVE",
      "items": [
        {
          "dataStoreRef": {
            "id": "ProvisionerDS",
            "location": "https://localhost:9999/pf-admin-api/v1/dataStores/ProvisionerDS"
          },
          "synchronizationFrequency": 60
        }
      ]
    }`

**Comment by ttranatping** [comment]
Once the docker image is released, I'd suggest also removing your outboundProvisioner config as it references ProvisioningDS

  "remove-config": [
    {
      "key": "id",
      "value": "ProvisionerDS",
      "parent": "items"
    },
    {
      "key": "resourceType",
      "value": "/serverSettings/outboundProvisioning"
    }
  ],

**Comment by PingDavidR** [comment]
Good morning @Jaylimboonkiat ... just to close this out, with the changes that @ttranatping put in place, you can follow the guide and build the Docker image locally using the provided Dockerfile under 99-helper-scripts/ping-bulkconfigtool.  We do not host the Docker image in our repository so there is no need to wait on us for that.  You can build it yourself as instructed in the guide after pulling in the changes.  

Thanks for bringing this to our attention!

**Comment by Jaylimboonkiat** [comment]
Thanks @ttranatping and @PingDavidR.

**Comment by samir-gandhi** [comment]
@Jaylimboonkiat I've opened a new ticket with the latest finding #424
